src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,683 @@
+#include "gfd_miner.h"
+
+#include <numeric>
+
+#include <boost/graph/copy.hpp>
+#include <boost/graph/isomorphism.hpp>
+#include <boost/graph/vf2_sub_graph_iso.hpp>
+#include <easylogging++.h>
+
+#include "config/option_using.h"
+
+namespace {
+
+using namespace algos;
+
+using Embedding = std::map<vertex_t, vertex_t>;
+using Embeddings = std::vector<Embedding>;
+
+using SingleRule = std::pair<std::vector<Literal>, Literal>;
+using Rule = std::pair<std::vector<Literal>, std::vector<Literal>>;
+using Rules = std::vector<Rule>;
+
+using Info = std::map<std::string, std::set<std::string>>;
+
+void NextSubset(std::vector<int>& indices, int const& border) {
+    if (indices.at(0) == border - static_cast<int>(indices.size())) {
+        indices = {};
+        return;
+    }
+    int index = 0;
+    for (int i = indices.size() - 1; i >= 0; --i) {
+        if (indices.at(i) != border - static_cast<int>(indices.size()) + i) {
+            index = i;
+            break;
+        }
+    }
+    indices.at(index)++;
+    for (int i = index + 1; i < static_cast<int>(indices.size()); ++i) {
+        indices.at(i) = indices.at(index) + i - index;
+    }
+}
+
+template <typename T>
+std::vector<std::vector<T>> GetSubsets(std::vector<T> const& elements, int const& n) {
+    if (static_cast<int>(elements.size()) < n) {
+        return {};
+    }
+    std::vector<std::vector<T>> result = {};
+    std::vector<int> indices(n);
+    std::iota(std::begin(indices), std::end(indices), 0);
+    while (!indices.empty()) {
+        std::vector<T> element = {};
+        for (auto index : indices) {
+            element.push_back(elements.at(index));
+        }
+        result.push_back(element);
+        NextSubset(indices, elements.size());
+    }
+    return result;
+}
+
+bool Satisfied(std::vector<Literal> const& literals, graph_t const& graph, graph_t const& pattern,
+               Embedding const& embedding) {
+    for (Literal const& l : literals) {
+        auto fst_token = l.first;
+        auto snd_token = l.second;
+        std::string fst;
+        std::string snd;
```

Comment:
Would perfectly fit on the same line without loss of readability
```suggestion
        std::string fst, snd;
```